addons/hr_holidays/models/hr_leave_type.py

```diff
@@ -107,48 +102,50 @@ def _search_valid(self, operator, value):
+        date_from = self._context.get('default_date_from') or fields.Date.today().strftime('%Y-1-1')
+        date_to = self._context.get('default_date_to') or fields.Date.today().strftime('%Y-12-31')
         employee_id = self._context.get('default_employee_id', self._context.get('employee_id')) or self.env.user.employee_id.id
 
         if not isinstance(value, bool):
             raise ValueError('Invalid value: %s' % (value))
         if operator not in ['=', '!=']:
             raise ValueError('Invalid operator: %s' % (operator))
-        new_operator = 'in' if operator == '=' else 'not in'
-
-        query = '''
-        SELECT
-            holiday_status_id
-        FROM
-            hr_leave_allocation alloc
-        WHERE
-            alloc.employee_id = %s AND
-            alloc.active = True AND alloc.state = 'validate' AND
-            (alloc.date_to >= %s OR alloc.date_to IS NULL) AND
-            alloc.date_from <= %s
-        '''
+        if (operator == '=') ^ value:
+            new_operator = 'not in'
+        else:
+            new_operator = 'in'
 
-        self._cr.execute(query, (employee_id or None, date_to, date_from))
+        leave_types = self.env['hr.leave.allocation'].search([
+            ('employee_id', '=', employee_id),
+            ('state', '=', 'validate'),
+            ('date_from', '<=', date_to),
+            '|',
+            ('date_to', '>=', date_from),
+            ('date_to', '=', False),
+        ]).holiday_status_id
 
-        return [('id', new_operator, [x['holiday_status_id'] for x in self._cr.dictfetchall()])]
+        return [('id', new_operator, leave_types.ids)]
 
-    @api.depends('requires_allocation')
+    @api.depends('requires_allocation', 'max_leaves', 'virtual_remaining_leaves')
     def _compute_valid(self):
-        date_to = self._context.get('default_date_to', fields.Datetime.today())
         date_from = self._context.get('default_date_from', fields.Datetime.today())
+        date_to = self._context.get('default_date_to', fields.Datetime.today())
         employee_id = self._context.get('default_employee_id', self._context.get('employee_id', self.env.user.employee_id.id))
         for holiday_type in self:
             if holiday_type.requires_allocation == 'yes':
-                allocation = self.env['hr.leave.allocation'].search([
+                allocations = self.env['hr.leave.allocation'].search([
                     ('holiday_status_id', '=', holiday_type.id),
                     ('employee_id', '=', employee_id),
+                    ('date_from', '<=', date_from),
                     '|',
                     ('date_to', '>=', date_to),
-                    '&',
                     ('date_to', '=', False),
-                    ('date_from', '<=', date_from)])
-                holiday_type.has_valid_allocation = bool(allocation)
+                ])
+                allocations = allocations.filtered(lambda alloc:
```

Comment:
because we rely on `max_leaves` and `virtual_remaining_leaves` that aren't stored, so we cannot look for them. I could add a search method for those but it feels overkill